uqcsbot/minecraft.py

```diff
@@ -0,0 +1,115 @@
+import logging
+import os
+from datetime import datetime
+
+import discord
+from aiomcrcon import Client, IncorrectPasswordError, RCONConnectionError
+from discord import app_commands, Colour
+from discord.ext import commands
+
+from uqcsbot.bot import UQCSBot
+from uqcsbot.models import MCWhitelist
+
+RCON_ADDRESS = os.environ.get("MC_RCON_ADDRESS")
+RCON_PORT = os.environ.get("MC_RCON_PORT")
+RCON_PASSWORD = os.environ.get("MC_RCON_PASSWORD")
+
+
+class Minecraft(commands.Cog):
+    def __init__(self, bot: UQCSBot):
+        self.bot = bot
+
+    async def send_rcon_command(self, command: str):
+        """ 
+        Sends a command via RCON to the server defined in environment variables.
+        
+        Args:
+        command: str - The command to send to the server.
+
+        Returns:
+        A tuple with the response message, and ID for the return message.
+        An ID of -1 is returned if there was an issue connecting to the server.
+        """
+        try:
+            async with Client(RCON_ADDRESS, RCON_PORT, RCON_PASSWORD) as client:
+                response = await client.send_cmd(command)
+
+        except RCONConnectionError:
+            return ("An error occured whilst connecting to the configured server.", -1)
+        except IncorrectPasswordError:
+            return ("The configured password is incorrect.", -1)
+
+        return response
+
+    @app_commands.command()
+    @app_commands.describe(username="Minecraft username to whitelist.")
+    async def mcwhitelist(self, interaction: discord.Interaction, username: str):
+        """ Adds a username to the whitelist for the UQCS server. """
+        db_session = self.bot.create_db_session()
+        query = db_session.query(MCWhitelist).filter(MCWhitelist.discord_id == interaction.user.id)
+        is_user_admin = interaction.user.guild_permissions.manage_guild
```

Comment:
There is a slight caveat to this, `interaction.user` can either be of type `User` or `Member`, and only `Member` has the `guild_permissions` field. I *think* this is to do with whether you send the bot a command in the server or not, as it won't be able to check your permissions without knowing which server you're talking about.

Reply:
Ah good catch, I'll put a check in for this at some point